### ISSUE TITLE
Json decoder not work when MillisecondsSinceEpoch represent as scientific notation

### ISSUE THREAD

**Issue by PChou** [trinodb/trino#7604]
trino version: <= 354
I'm testing kafka connector, the source topic data show as below(use kafka console consumer): 

```json
{"traceId":"06e52ab0-2b96-4cf1-a92c-0e64ee385d16", "timestamp":1.618540600943E12}
```
notice that, timestamp is represent milliseconds since epoch, but show as scientific notation.
Then I try to configure the table descriptor

```json
{
"message": {
    "dataFormat": "json",
    "fields": [{
      "name": "traceId",
      "type": "VARCHAR",
      "mapping": "traceId"
    },{
      "name": "timestamp",
      "type": "TIMESTAMP",
      "mapping": "timestamp",
      "dataFormat": "milliseconds-since-epoch"
    }]
  }
}
```

but the query failed: 

```
Query failed: could not parse value '1.618458216014E12' as 'timestamp(3)' for column 'timestamp'
```

from `io/trino/decoder/json/MillisecondsSinceEpochJsonFieldDecoder.java` I found the exception root cause: 

```
protected long getMillis()
        {
            if (value.isIntegralNumber() && !value.isBigInteger()) {
                return value.longValue();
            }
            if (value.isValueNode()) {
                try {
                    return parseLong(value.asText());
                }
                catch (NumberFormatException e) {
                    throw new TrinoException(
                            DECODER_CONVERSION_NOT_SUPPORTED,
                            format("could not parse value '%s' as '%s' for column '%s'", value.asText(), columnHandle.getType(), columnHandle.getName()));
                }
            }
            throw new TrinoException(
                    DECODER_CONVERSION_NOT_SUPPORTED,
                    format("could not parse non-value node as '%s' for column '%s'", columnHandle.getType(), columnHandle.getName()));
        }
```

Failure happens at  parseLong("1.618458216014E12")

Since some json framework(ie. jackson) may encode long into the format of scientific notation, do we need to support this case?

**Comment by hashhar** [comment]
I think the issue is that we are treating a JSON Number as an integer/long value when it in fact is a floating point value.

I think this should be as easy as reading the value into a double and casting back into long instead (while being aware of when the value is non-integral).

**Comment by JoJoBizarreAdventure** [comment]
I found that it is parsed to be a **FloatNode**, which should be regarded as **invalid** in **TestMillisecondsSinceEpochJsonFieldDecoder**.

**Comment by PChou** [comment]
@hashhar @JoJoBizarreAdventure 
If floating-point numbers are supported, there is indeed a loss of precision. We need to decide how to discard units smaller than milliseconds.

**Comment by hashhar** [comment]
@PChou IMO:

1. Use `Double.parseDouble` to read the values from JSON since JSON doesn't have an integer type.
2. If the parsed value has a fractional part - throw an exception because none of the formatters support fractional parts of seconds/millis/micros. Silent truncation can lead to misleading expected and actual values for the end-user.

Not that not truncating also has the benefit that it matches the behaviour we already have today.

cc: @losipiuk WDYT?

**Comment by losipiuk** [comment]
> Not that not truncating also has the benefit that it matches the behaviour we already have today.

I would go with truncating. There is a chance that someone depends on support of non 0 fractional part right now.

**Comment by hashhar** [comment]
> > Not that not truncating also has the benefit that it matches the behaviour we already have today.
> 
> I would go with truncating. There is a chance that someone depends on support of non 0 fractional part right now.

Unlikely because in current code it would throw a `NumberFormatException` due to the `parseLong` call.

**Comment by losipiuk** [comment]
> > > Not that not truncating also has the benefit that it matches the behaviour we already have today.
> > 
> > 
> > I would go with truncating. There is a chance that someone depends on support of non 0 fractional part right now.
> 
> Unlikely because in current code it would throw a `NumberFormatException` due to the `parseLong` call.

Yeah, you are right. Then validation seems fine give any json value representing integer should be represented as integer `double` value (without fractional part)  under IEEE 754.
